File: 2022/14/14.py
from re import findall


def preprocessing(input): 
    paths = []
    max_depth = 0
    rocks = set()
    for path in input.splitlines():
        numbers = list(map(int, findall(r'[0-9]+',path)))
        xa, ya = numbers.pop(0), numbers.pop(0)
        while numbers:
            xa, ya, xb, yb = numbers.pop(0), numbers.pop(0), xa, ya
            if xa == xb: 
                for dy in range(min(ya, yb), max(ya, yb) + 1): rocks.add((xa, dy))
            else: 
                for dx in range(min(xa, xb), max(xa, xb) + 1): rocks.add((dx, ya))
                
    max_depth =  max(y for _, y in rocks) + 2
    for x in range(1000): rocks.add((x, max_depth))
    return rocks

    
def solver(rocks):
    x, y, sand = 500, 0, 0
    endless = True
    max_depth = max(y for _, y in rocks)
    while (500, 0) not in rocks:
        if   (x    , y + 1) not in rocks : y += 1
        elif (x - 1, y + 1) not in rocks : (x , y) = (x - 1, y + 1)
        elif (x + 1, y + 1) not in rocks : (x , y) = (x + 1, y + 1)
        else :
            rocks.add((x, y))
            (x, y) = (500, 0)
            sand   += 1
        if endless and y > max_depth - 2: 
            yield sand
            endless = False
    yield sand

File: 2022/14/14.rs
aoc2022::main!() ;

fn generator(input : &str) -> ([[u32 ; 1000] ; 200] , usize) {
    let mut cave = [[0 ; 1000] ; 200];
    let mut max_depth = 0 ;
    input
        .lines()
        .map(|l| l.split(" -> ").map(|c| c.split(',').map(|n| n.parse::<usize>().unwrap()).collect_vec()).collect_vec())
        .for_each(|path| path.iter().tuple_windows().for_each(|(start, end)| 
            iproduct!(start[0].min(end[0]) .. start[0].max(end[0]) + 1 , start[1].min(end[1]) .. start[1].max(end[1]) + 1)
                .for_each(|(x, y)| {
                max_depth = max_depth.max(y) ; 
                cave[y][x]=1 ; })
    ));
    ( cave, max_depth)
}

fn part_1(input :  ([[u32 ; 1000] ; 200] , usize)) -> u32 {
    solver(input.0, input.1, false)
}

fn part_2(input :  ([[u32 ; 1000] ; 200] , usize)) -> u32 {
    solver(input.0, input.1 , true)
}

fn solver(mut cave : [[u32 ; 1000] ; 200], mut max_depth : usize, floor : bool) -> u32 {

    let mut sand = 0 ; 
    let (mut x, mut y) = (500 ,0);

    if floor { 
        max_depth += 2 ;
        (0 .. 1000).for_each(|x| cave[max_depth][x] = 1)} ;
        
    loop {
        if cave[0][500] == 1 || y >= max_depth {return sand} 
        else if cave[y+1][x] == 0  {y += 1}
        else if cave[y+1][x-1] == 1 && cave[y+1][x+1] == 1 {
            cave[y][x] = 1 ;
            sand += 1 ;
            (x, y) = (500, 0)  ; 
        } 
        else if cave[y+1][x-1] == 0 { (x , y) = (x - 1, y + 1) }
        else if cave[y+1][x+1] == 0 { (x , y) = (x + 1, y + 1) }
    }
    }